File: frontend/src/types/graph.ts
export type ConfidenceLevel = 'EXTRACTED' | 'INFERRED' | 'AMBIGUOUS'

export interface KnowledgeNode {
  id: string
  name: string
  description: string
  category: string
  color: string
  size: number
  source_id: number
  source_type: string
  community_id: number | null
  mastery: number
}

export interface KnowledgeLink {
  id: number
  source_node_id: string
  target_node_id: string
  relationship: string
  confidence: ConfidenceLevel
  confidence_score: number
  weight: number
}

export interface Hyperedge {
  id: number
  label: string
  relation: string
  confidence: ConfidenceLevel
  confidence_score: number
  node_ids: string[]
}

export interface GraphData {
  nodes: KnowledgeNode[]
  edges: KnowledgeLink[]
  hyperedges: Hyperedge[]
}

File: backend/app/schemas/graph.py
from pydantic import BaseModel


class NodeOut(BaseModel):
    id: str
    name: str
    description: str
    category: str
    color: str
    size: float
    source_id: int
    source_type: str
    community_id: int | None = None
    mastery: float

    model_config = {"from_attributes": True}


class EdgeOut(BaseModel):
    id: int
    source_node_id: str
    target_node_id: str
    relationship: str
    confidence: str
    confidence_score: float
    weight: float

    model_config = {"from_attributes": True}


class HyperedgeOut(BaseModel):
    id: int
    label: str
    relation: str
    confidence: str
    confidence_score: float
    node_ids: list[str]

    model_config = {"from_attributes": True}


class GraphOut(BaseModel):
    nodes: list[NodeOut]
    edges: list[EdgeOut]
    hyperedges: list[HyperedgeOut]
